README.md
# libpca9506

C driver for the NXP Semiconductors PCA9506/06 40-bit I2C-bus I/O port with RESET, OE and INT ([datasheet](https://github.com/carlodicelico/libpca9506/tree/master/docs/PCA9506_06.pdf))

## Features

- Tests with [greatest](https://github.com/silentbicycle/greatest) and [theft](https://github.com/silentbicycle/theft)
- Bring your own I2C and hardware interface libraries
- Minimal dependencies

## Goals

- Strict, full implementation of manufacturer specifications
- A clean, user-friendly interface for application developers and library writers
- No tight coupling to a specific I2C library or hardware platform (Linux I2C, various Pi/Arduino libraries, etc.)
- No tight coupling to a specific application (LEDs vs. servos, etc.)
- Modern C (C11) development practices, like unit tests, property-based tests, [modern CMake (3.13)](https://cliutils.gitlab.io/modern-cmake/), etc.

## Installation

**Option 1:** add it as a git submodule like this:

```shell
$ git clone git@github.com:carlodicelico/libpca9506.git vendor/libpca9506
```

Then, add it in your `CMakeLists.txt` and link your library or executable:
```cmake
add_subdirectory(vendor/libpca9506)
target_link_libraries(my_lib pca9506)
```

**Option 2:** install it like this:

```shell
$ git clone git@github.com:carlodicelico/libpca9506.git && cd libpca9506
$ cmake -DCMAKE_BUILD_TYPE=Release -DTESTING=OFF -G "Unix Makefiles" -S . -B cmake-build-release
$ cd cmake-build-release
$ make && sudo make install 
```

Then, add it in your `CMakeLists.txt` and link your library or executable:

```cmake
find_library(PCA9506 pca9506)
target_link_libraries(my_lib ${PCA9506})
```

## Usage

Example application code might look something like this:

```C
#include "libpca9506.h"

// Application (not library) dependencies
#include "pigpio.h"
#include <zconf.h>

#define I2C_BUS 1
#define I2C_BUS_ADDRESS 0x40
#define I2C_FLAGS 0

// ALL management of the I2C bus itself is your application's responsibility!
static int pi, handle;

// Set up an I2C reader callback
static uint8_t i2c_read_byte_cb(pca9506_s *driver, uint8_t r) {
    // This is pigpio's function we're wrapping here, but it could be any other library, as well
    int result = i2c_read_byte_data(pi, (uint8_t)servo_master_handle, r);

    // You can add information to the driver struct's status field if you like
    driver->status = (result >= 0) ? "ok" : pigpio_error(result);

    return (uint8_t)result;
}

static uint8_t i2c_write_byte_cb(pca9506_s *driver, uint8_t r, uint8_t d) {
    // This is pigpio's function we're wrapping here, but it could be any other library, as well
    int result = i2c_write_byte_data(pi, (uint8_t)servo_master_handle, r, d);

    // You can add information to the driver struct's status field if you like
    driver->status = (result >= 0) ? "ok" : pigpio_error(result);

    return (uint8_t)result;
}

int main(int argc, char **argv) {
    pi = pigpio_start(NULL, NULL);
    char *pi_status = (pi >= 0) ? "ok" : pigpio_error(pi);

    handle = i2c_open(pi, I2C_BUS, I2C_BUS_ADDRESS, I2C_FLAGS);
    char *handle_status = (handle >= 0) ? "ok" : pigpio_error(handle);

    printf("\nMain\n====\n: %d");
    printf("\nPIGPIO Status: %s", pi_status);
    printf("\nI2C Handle Status: %s\n", handle_status);
    printf("Initializing driver...");

    // Initialize your driver struct—also does test read and write operations
    pca9506_s my_driver = pca9506(.bus_reader=bus_byte_reader, .bus_writer=bus_byte_writer);

    printf("%s\n", my_driver.status);

    // set outputs
    my_driver.clear_bank(&my_driver, IOC0);
    my_driver.clear_bank(&my_driver, IOC1);
    my_driver.clear_bank(&my_driver, IOC2);
    my_driver.clear_bank(&my_driver, IOC3);
    my_driver.clear_bank(&my_driver, IOC4);

    // read some data from IP0
    my_driver.read_bank(&my_driver, IP0);

    // set 3rd bit in OP3
    my_driver.set_bit_in_bank(&my_driver, OP3, 2);

    // etc..

    // pigpio-specific...
    int close = i2c_close(pi, (uint)handle);
    char *close_message = (close == PI_BAD_HANDLE) ? "PI_BAD_HANDLE" : "CLOSED";
    printf("\nClose result: %s\n", close_message);

    pigpio_stop(pi);

    if(close == PI_BAD_HANDLE) return 1;

    return 0;
}
```

See also the included [examples](https://github.com/carlodicelico/libpca9506/tree/master/examples).

## How do I contribute?

Contributions are welcome! Check out [CONTRIBUTING](https://github.com/carlodicelico/libpca9506/tree/master/CONTRIBUTING.md) for instructions.

## How is this library versioned?

This library follows the principles of [Semantic Versioning](http://semver.org/). You can find each new release,
along with the changelog, in the [CHANGELOG](https://github.com/carlodicelico/libpca9506/tree/master/CHANGELOG.md).

During initial development, the major version will be 0 (e.g., `0.x.y`), which indicates the code does not yet have a
stable API. Once we hit `1.0.0`, we will make every effort to maintain a backwards compatible API and use the MAJOR,
MINOR, and PATCH versions on each release to indicate any incompatibilities.

## License

This code is released under the MIT License. Please see [LICENSE](https://github.com/carlodicelico/libpca9506/tree/master/LICENSE) for more details.



CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](https://keepachangelog.com/en/1.0.0/),
and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0.html).

## [Unreleased]
- TBD

## v0.2.1
- Make installable
## v0.2.0
### Added
- Project setup, including:
  - CMake setup
  - Unit and property-based tests
- Driver library, including:
  - Interface for user-provided I2C read/write callbacks
  - Basic data structure for the driver
  - Definitions for key manufacturer register addresses and default values
- Public API, including:
  - Reading, setting, and clearing register bits
  - Reading, setting, and clearing bank byte data
- Documentation, including:
  - this CHANGELOG
  - [AUTHORS](https://github.com/carlodicelico/libpca9506/tree/master/AUTHORS)
  - [CONDUCT](https://github.com/carlodicelico/libpca9506/tree/master/CONDUCT.md)
  - [CONTRIBUTING](https://github.com/carlodicelico/libpca9506/tree/master/CONTRIBUTING.md)
  - [LICENSE](https://github.com/carlodicelico/libpca9506/tree/master/LICENSE)
  - [README](https://github.com/carlodicelico/libpca9506/tree/master/README.md)


tests/runner.c
#include "runner.h"

GREATEST_MAIN_DEFS();

int main(int argc, char **argv) {
    (void)argc;
    (void)argv;

    struct greatest_report_t report;

    GREATEST_INIT();
    greatest_set_verbosity(3);

    RUN_SUITE(test_driver_init);
    RUN_SUITE(test_register_ops);

    GREATEST_PRINT_REPORT();

    greatest_get_report(&report);

    if(report.failed > 0) return 1;

    return 0;
}


tests/runner.h
#ifndef RUNNER_H
#define RUNNER_H

#include "suites.h"

#endif


tests/suites/tests.c
#include "tests.h"

#include <stdint.h>

mock_register_s mockr;

void set_up_mock_driver(pca9506_s *driver) {
    *driver = pca9506(.bus_reader=mock_bus_reader, .bus_writer=mock_bus_writer);
}

void set_up_theft_run_config(struct theft_run_config *config) {
    theft_seed seed = theft_seed_of_time();

    *config = (struct theft_run_config){
        .type_info = {
            theft_get_builtin_type_info(THEFT_BUILTIN_uint8_t)
        },
        .seed = seed,
        .trials = 100
    };
}

uint8_t mock_bus_reader(pca9506_s *driver, uint8_t address) {
    (void)driver;

    mockr.address = address;
    return mockr.value;
}

uint8_t mock_bus_writer(pca9506_s *driver, uint8_t address, uint8_t data_in) {
    (void)driver;

    mockr.address = address;
    mockr.value = data_in;

    return 0;
}



src/pca9506.c
#include "pca9506.h"
#include "registers.h"

#include <stdio.h>

// A convenience... you could also get the bit from the byte in driver->data
static uint read_bit_from_bank(pca9506_s *h, u8 bank_address, uint bit_position) {
    return get_bit(h, bank_address, bit_position);
}

static void set_bit_in_bank(pca9506_s *h, u8 bank_address, uint bit_position) {
    set_bit(h, bank_address, bit_position);
}

static void clear_bit_in_bank(pca9506_s *h, u8 bank_address, uint bit_position) {
    clear_bit(h, bank_address, bit_position);
}

static void read_bank(pca9506_s *h, u8 bank_address) {
    pca9506_i2c_bus_read(h, bank_address);
}

static void set_bank(pca9506_s *h, u8 bank_address) {
    pca9506_i2c_bus_write(h, bank_address, HIGH);
}

static void clear_bank(pca9506_s *h, u8 bank_address) {
   pca9506_i2c_bus_write(h, bank_address, LOW);
}

pca9506_s pca9506_configure_handle(pca9506_s driver){
    if(!(driver.bus_reader && driver.bus_writer)){
        driver.command = "cb_check";
        driver.status = "Invalid input: bus reader and bus writer callbacks are both required.";
    } else {
        driver.command = "rw_check";

        // Quick & dirty read/write checks; only verifies the interface, should break on the caller's side.
        pca9506_i2c_bus_read(&driver, 0x18);
        pca9506_i2c_bus_write(&driver, 0x18, driver.data);
    }

    driver.read_bit_from_bank = read_bit_from_bank;
    driver.set_bit_in_bank = set_bit_in_bank;
    driver.clear_bit_in_bank = clear_bit_in_bank;
    driver.read_bank = read_bank;
    driver.set_bank = set_bank;
    driver.clear_bank = clear_bank;

    driver.command = driver.command == NULL ? "init" : driver.command;
    driver.status = driver.status == NULL ? "ok" : driver.status;

    return driver;
}



src/registers.c
#include "registers.h"

/** Userland I2C bus read/write callback wrappers */

void pca9506_i2c_bus_read(pca9506_s *h, u8 r) {
    u8 result = h->bus_reader(h, r);
    h->command = "i2c_read";
    h->status = "ok";
    h->address = r;
    h->data = result;
}

void pca9506_i2c_bus_write(pca9506_s *h, u8 r, u8 d) {
    u8 result = h->bus_writer(h, r, d);
    h->command = "i2c_write";
    h->status = result == 0 ? "ok" : "error";
    h->address = r;
    h->data = d;
}

void set_bit(pca9506_s *h, u8 r, uint b) {
    pca9506_i2c_bus_read(h, r);
    pca9506_i2c_bus_write(h, r, SET_BIT(h->data, b));
}

void clear_bit(pca9506_s *h, u8 r, uint b) {
    pca9506_i2c_bus_read(h, r);
    pca9506_i2c_bus_write(h, r, CLEAR_BIT(h->data, b));
}

uint get_bit(pca9506_s *h, u8 r, uint b) {
    pca9506_i2c_bus_read(h, r);
    return GET_BIT(h->data, b);
}



src/libpca9506.c
#include "libpca9506.h"
#include "registers.h"

#include <stdio.h>

// A convenience... you could also get the bit from the byte in driver->data
static uint read_bit_from_bank(pca9506_s *h, u8 bank_address, uint bit_position) {
    return get_bit(h, bank_address, bit_position);
}

static void set_bit_in_bank(pca9506_s *h, u8 bank_address, uint bit_position) {
    set_bit(h, bank_address, bit_position);
}

static void clear_bit_in_bank(pca9506_s *h, u8 bank_address, uint bit_position) {
    clear_bit(h, bank_address, bit_position);
}

static void read_bank(pca9506_s *h, u8 bank_address) {
    pca9506_i2c_bus_read(h, bank_address);
}

static void set_bank(pca9506_s *h, u8 bank_address) {
    pca9506_i2c_bus_write(h, bank_address, HIGH);
}

static void clear_bank(pca9506_s *h, u8 bank_address) {
   pca9506_i2c_bus_write(h, bank_address, LOW);
}

pca9506_s pca9506_configure_handle(pca9506_s driver){
    if(!(driver.bus_reader && driver.bus_writer)){
        driver.command = "cb_check";
        driver.status = "Invalid input: bus reader and bus writer callbacks are both required.";
    } else {
        driver.command = "rw_check";

        // Quick & dirty read/write checks; only verifies the interface, should break on the caller's side.
        pca9506_i2c_bus_read(&driver, 0x18);
        pca9506_i2c_bus_write(&driver, 0x18, driver.data);
    }

    driver.read_bit_from_bank = read_bit_from_bank;
    driver.set_bit_in_bank = set_bit_in_bank;
    driver.clear_bit_in_bank = clear_bit_in_bank;
    driver.read_bank = read_bank;
    driver.set_bank = set_bank;
    driver.clear_bank = clear_bank;

    driver.command = driver.command == NULL ? "init" : driver.command;
    driver.status = driver.status == NULL ? "ok" : driver.status;

    return driver;
}



src/libpca9606.c
#include "libpca9506.h"
#include "registers.h"

#include <stdio.h>

pca9506_s pca9506_configure_handle(pca9506_s driver){
    if(!(driver.bus_reader && driver.bus_writer)){
        driver.command = "cb_check";
        driver.status = "Invalid input: bus reader and bus writer callbacks are both required.";
    } else {
        driver.command = "rw_check";

        // Quick & dirty read/write checks; only verifies the interface, should break on the caller's side.
        pca9506_i2c_bus_read(&driver, 0x18);
        pca9506_i2c_bus_write(&driver, 0x18, driver.data);
    }

    driver.command = driver.command == NULL ? "init" : driver.command;
    driver.status = driver.status == NULL ? "ok" : driver.status;

    return driver;
}

/** Get/set I/O mode for the given pin(s) */

void set_as_output(pca9506_s *h, int p) {
    set_register_byte(h, IOC_REG, (uint8_t)p, 0xFF);
}

void set_as_input(pca9506_s *h, int p) {
    set_register_byte(h, IOC_REG, (uint8_t)p, 0x00);
}

void set_as_outputs(pca9506_s *h) {
    set_register_bytes(h, IOC_REG, 0xFF);
}

void set_as_inputs(pca9506_s *h) {
    set_register_bytes(h, IOC_REG, 0x00);
}

void io_read(pca9506_s *h, int p) {
    get_register_byte(h, IP_REG, (uint8_t)p);
}

void io_write(pca9506_s *h, int p, uint8_t v) {
    set_register_byte(h, OP_REG, (uint8_t)p, v);
}

void set_high(pca9506_s *h) {
    set_register_bytes(h, OP_REG, 0xFF);
}

void set_low(pca9506_s *h) {
    set_register_bytes(h, OP_REG, 0x00);
}



include/registers.h
#ifndef PCA9506_REGISTERS_H
#define PCA9506_REGISTERS_H

#include "pca9506.h"

#include <inttypes.h>

/** Make static functions available for testing */
#ifdef TESTING
#define STATIC
#else
#define STATIC static
#endif

/** Bitwise operations */
#define GET_MASK(n) (1u << (n))                 // Get the bitmask for a given number n
#define GET_BIT(n, b) ((n >> b) & 1u)           // Get the bit at b from number n
#define SET_BIT(n, b) ((n) |= (1u << (b)))      // Set the bit b of number n
#define CLEAR_BIT(n, b) ((n) &= ~(1u << (b)))   // Clear the bit b of number n

/** Format binary for logging/testing, e.g.,
 *  printf("Data in binary: "BYTE_TO_BIN"\n", TO_BINARY(d));
 */
#define BYTE_TO_BIN "%c%c%c%c%c%c%c%c"
#define TO_BINARY(byte)  \
  (byte & 0x80 ? '1' : '0'), \
  (byte & 0x40 ? '1' : '0'), \
  (byte & 0x20 ? '1' : '0'), \
  (byte & 0x10 ? '1' : '0'), \
  (byte & 0x08 ? '1' : '0'), \
  (byte & 0x04 ? '1' : '0'), \
  (byte & 0x02 ? '1' : '0'), \
  (byte & 0x01 ? '1' : '0')

/** Conveniences */
#define ZERO 0x00
#define OK   ZERO

#define ONE  0x01
#define ERR  ONE

/** Low-level access to user callback wrappers for initialization */
void pca9506_i2c_bus_read(pca9506_s *h, u8 r);
void pca9506_i2c_bus_write(pca9506_s *h, u8 r, u8 d);

/** Private API */

void set_bit(pca9506_s *h, u8 r, uint b);
void clear_bit(pca9506_s *h, u8 r, uint b);
uint get_bit(pca9506_s *h, u8 r, uint b);

#endif



include/libpca9606.h
/**
 * \file libpca9606.h
 *
 * \brief Driver for the NXP Semiconductors PCA9605/06 40-bit I2C-bus I/O port with RESET, OE and INT
 *
 * \see https://www.nxp.com/docs/en/data-sheet/PCA9505_9506.pdf
 * \see https://github.com/torvalds/linux/blob/master/include/linux/i2c.h#L152
 */

#ifndef PCA9606_LIBRARY_H
#define PCA9606_LIBRARY_H

#include <inttypes.h>

/** Initializes a driver_handle struct with values provided by the user. */
#define pca9606(...) (pca9606_configure_handle((pca9606_s){__VA_ARGS__}))

/** Primitive data type aliases for readability. */
typedef const char *command, *status;

/** Aggregate types */
struct pca9606_driver_handle;

/**
 * Types for the I2C bus read and write callbacks.
 * Modeled on the I2C interface for Linux (\c i2c_smbus_read_byte_data, etc.)
 *
 * \see https://github.com/torvalds/linux/blob/master/include/linux/i2c.h#L152
 */
typedef uint8_t (*pca9606_i2c_bus_read_cb)(struct pca9606_driver_handle *, uint8_t);
typedef uint8_t (*pca9606_i2c_bus_write_cb)(struct pca9606_driver_handle *, uint8_t , uint8_t);

/** Type definition for the driver handle */
typedef struct pca9606_driver_handle {
    uint8_t data;                  // Data last read/written
    uint8_t address;               // Last register read to/written from
    command command;               // Last command attempted
    status status;                 // Status (ok) or error message
    pca9606_i2c_bus_read_cb bus_reader;    // I2C Bus reader callback
    pca9606_i2c_bus_write_cb bus_writer;   // I2C Bus writer callback
} pca9606_s;

/** Private, used by the macro defined above. */
pca9606_s pca9606_configure_handle(pca9606_s);

/** Public API */

/** Set the given pin as an I/O output */
void set_as_output(pca9606_s *h, int pin);

/** Set the given pin as an I/O input */
void set_as_input(pca9606_s *h, int pin);

/** Set all pins as I/O outputs */
void set_as_outputs(pca9606_s *h);

/** Set all pins as I/O inputs */
void set_as_inputs(pca9606_s *h);

/** Read I/O data */
void io_read(pca9606_s *h, int pin);

/** Write I/O data */
void io_write(pca9606_s *h, int pin, uint8_t value);

/** Set output pins high */
void set_high(pca9606_s *h);

/** Set output pins low */
void set_low(pca9606_s *h);

#endif


include/libpca9506.h
/**
 * \file libpca9506.h
 *
 * \brief Driver for the NXP Semiconductors PCA9605/06 40-bit I2C-bus I/O port with RESET, OE and INT
 *
 * \see https://www.nxp.com/docs/en/data-sheet/PCA9505_9506.pdf
 * \see https://github.com/torvalds/linux/blob/master/include/linux/i2c.h#L152
 */

#ifndef PCA9506_LIBRARY_H
#define PCA9506_LIBRARY_H

#include <inttypes.h>

// Register definitions—pp. 9–10

/** Input port registers */
#define IP0 0x00
#define IP1 0x01
#define IP2 0x02
#define IP3 0x03
#define IP4 0x04

/** Output port registers */
#define OP0 0x08
#define OP1 0x09
#define OP2 0x0A
#define OP3 0x0B
#define OP4 0x0C

/** Polarity inversion registers */
#define PI0 0x10
#define PI1 0x11
#define PI2 0x12
#define PI3 0x13
#define PI4 0x14

/** I/O Configuration registers */
#define IOC0 0x18
#define IOC1 0x19
#define IOC2 0x1A
#define IOC3 0x1B
#define IOC4 0x1C

/** Mask interrupt registers */
#define MSK0 0x20
#define MSK1 0x21
#define MSK2 0x22
#define MSK3 0x23
#define MSK4 0x24

/** Bit positions—last 3 bits are reserved */
#define D0 0
#define D1 1
#define D2 2
#define D3 3
#define D4 4

/** Bit values */

#define HIGH 0xFF
#define LOW  0x00

/** Initializes a driver_handle struct with values provided by the user. */
#define pca9506(...) (pca9506_configure_handle((pca9506_s){__VA_ARGS__}))

/** Primitive data type aliases for readability/conciseness */
typedef char *string;
typedef uint8_t u8;
typedef unsigned int uint;

/** Aggregate types */

struct pca9506_driver;

/**
 * Types for the I2C bus read and write callbacks.
 * Modeled on the I2C interface for Linux (\c i2c_smbus_read_byte_data, etc.)
 *
 * \see https://github.com/torvalds/linux/blob/master/include/linux/i2c.h#L152
 */
typedef u8 (*pca9506_i2c_bus_read_cb)(struct pca9506_driver *driver, u8 address);
typedef u8 (*pca9506_i2c_bus_write_cb)(struct pca9506_driver *driver, u8 address, u8 data);

/** Public API function signatures */
typedef void (*pca9506_fn)(struct pca9506_driver *driver);
typedef uint (*pca9506_bit_read)(struct pca9506_driver *driver, u8 bank_address, uint bit_position);
typedef void (*pca9506_bit_op)(struct pca9506_driver *driver, u8 bank_address, uint bit_position);
typedef void (*pca9506_bank_op)(struct pca9506_driver *driver, u8 bank_address);

/** Type definition for the driver handle */
typedef struct pca9506_driver {
    u8 data;                                // Data last read/written
    u8 address;                             // Last register read to/written from
    string command;                         // Last command attempted
    string status;                          // Status (ok) or error message
    pca9506_i2c_bus_read_cb bus_reader;     // I2C Bus reader callback
    pca9506_i2c_bus_write_cb bus_writer;    // I2C Bus writer callback
    pca9506_bit_read read_bit_from_bank;    // Read the bit value at position p from bank b
    pca9506_bit_op set_bit_in_bank;         // Set the bit (1) at position p in bank b
    pca9506_bit_op clear_bit_in_bank;       // Clear the bit (0) at position p in bank b
    pca9506_bank_op read_bank;              // Read byte data from bank into handle.data
    pca9506_bank_op set_bank;               // Set ALL bits (1) in bank
    pca9506_bank_op clear_bank;             // Clear ALL bits (0) in bank
} pca9506_s;

/** Private, used by the macro defined above. */
pca9506_s pca9506_configure_handle(pca9506_s);

#endif
